application/controllers/OptinController.php

```diff
@@ -27,68 +27,199 @@ class OptinController extends LSYii_Controller
+        $languageCode = Yii::app()->request->getQuery('langcode');
+        $accessToken = Token::sanitizeToken(Yii::app()->request->getQuery('token'));
+
+        //IF there is no survey id, redirect back to the default public page
+        if (!$surveyId) {
+            $this->redirect(['/']);
+        }
+
+        $survey = Survey::model()->findByPk($surveyId);
+        if (empty($survey) || !$survey->hasTokensTable) {
+            throw new CHttpException(404, "This survey does not seem to exist. It may have been deleted or the link you were given is outdated or incorrect.");
+        }
+
+        //Check that there is a SID
+        // Get passed language from form, so that we dont lose this!
+        if (!isset($languageCode) || $languageCode == "" || !$languageCode) {
+            $baseLanguage = $survey->language;
+        } else {
+            $baseLanguage = sanitize_languagecode($languageCode);
+        }
+
+        Yii::app()->setLanguage($baseLanguage);
+
+        LimeExpressionManager::singleton()->loadTokenInformation($surveyId, $accessToken, false);
+        $token = Token::model($surveyId)->findByAttributes(['token' => $accessToken]);
+
+        $link = '';
+        if (!isset($token)) {
+            $message = gT('You are not a participant of this survey.');
         } else {
-            $sBaseLanguage = sanitize_languagecode($sLanguageCode);
+            $optedOutFromSurvey = substr($token->emailstatus, 0, strlen('OptOut')) == 'OptOut';
+
+            $blacklistHandler = new LimeSurvey\Models\Services\ParticipantBlacklistHandler();
+            $participant = $blacklistHandler->getCentralParticipantFromToken($token);
+            $isBlacklisted = !empty($participant) && $participant->blacklisted == 'Y';
+
+            if (!Yii::app()->getConfig('allowunblacklist') == "Y") {
+                $message = gT('Removing yourself from the blacklist is currently disabled.');
+            } elseif ($isBlacklisted) {
+                $message = gT('Please confirm that you want to be added back to the central participants list for this site.');
+                $link = Yii::app()->createUrl('optin/addtokens', ['surveyid' => $surveyId, 'langcode' => $baseLanguage, 'token' => $accessToken, 'global' => true]);
+            } elseif ($optedOutFromSurvey) {
+                $message = gT('Please confirm that you want to be added back to this survey by clicking the button below.') . '<br>' . gT("After confirmation you may start receiving invitations and reminders for this survey.");
+                $link = Yii::app()->createUrl('optin/addtokens', ['surveyid' => $surveyId, 'langcode' => $baseLanguage, 'token' => $accessToken]);
+            } elseif (empty($participant)) {
+                $message = gT('You are already a participant of this survey.');
+            } else {
+                $message = gT('You are already part of the central participants list for this site.');
+            }
```

Comment:
Can we send data about situation in view ?

Something like `['participant' => $participant->getAttributes(), 'token' => $token->getAttributes()]` etc …

Else : twig edition don't know anything on the reason  , best is to put gT in twig but take more time.

Reply:
> Can we send data about situation in view ?

Why?

Reply:
To allow Survey theme to do different job according to situation. And show own string with own class and action …

Reply:
Find a better solution than br in https://github.com/LimeSurvey/LimeSurvey/pull/2202/files#diff-b5c799a27b238495f0849ffe1a47677968dde9b20817563805cb573c1b031fccR131 

br tag is clearly a view … ;)

Alternative solution : participant.attribute and token.attribute : you say it's logic, i say it's between ;).

I'm sure someone ask/need a solution to do something according to «situation» in this page. Have a solution nee to have information ;).